Duplicate symbol link error in 3.1.48 when including bind.h in more than one source file.  Ok in 3.1.46.

I am compiling my projects using the Docker images [here](https://hub.docker.com/r/emscripten/emsdk), as referenced in [the docs](https://emscripten.org/docs/getting_started/downloads.html#using-the-docker-image).  I'm running Docker Desktop on Windows 11(WSL2).

Below is the output of my build, first under 3.1.46, then under 3.1.48.

Following that are the sources I created to reproduce this, and have attached a zip file with all files, including a makefile and verbose build logs.  You can run 'make clean' to clear the cache, and 'make VERBOSE=true' to set EMCC_DEBUG := 2.  And 'make TYPE=opt' will do an optimized build.
[linkerr.zip](https://github.com/emscripten-core/emscripten/files/13301658/linkerr.zip)

I've included the makefile to show clearly the use all of the flags we're using in our project - I've tried removing all of these except -std=c++20 and -v, and 3.1.48 still gives the link error.

### Build under 3.1.46:
```
root@a8c172869196:/mnt/linkerr# emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.46 (19607820c447a13fd8d0b7680c56148427d6e1b8)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

root@a8c172869196:/mnt/linkerr# 
root@a8c172869196:/mnt/linkerr# make
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file1.o file1.cpp
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file2.o file2.cpp
  Linking 
 "/emsdk/upstream/bin/clang" --version
cache:INFO: generating system asset: symbol_lists/0193838d629247ad34612c7ac6c8b5797e3de0b2.json... (this will be cached in "/emsdk/upstream/emscripten/cache/symbol_lists/0193838d629247ad34612c7ac6c8b5797e3de0b2.json" for subsequent builds)
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/src/compiler.js /tmp/tmp3_jj52we.json --symbols-only
cache:INFO:  - ok
 "/emsdk/upstream/bin/wasm-ld" -o out.wasm ./file1.o ./file2.o -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten --whole-archive -lembind-rtti --no-whole-archive -lGL-webgl2-full_es3 -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc-debug -lcompiler_rt-wasm-sjlj -lc++-except -lc++abi-debug-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm /tmp/tmpp9lat0t3libemscripten_js_symbols.so --export-if-defined=main --export-if-defined=__get_exception_message --export-if-defined=free --export-if-defined=__cpp_exception --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__thrown_object_from_unwind_exception --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=sbrk --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=emscripten_get_sbrk_ptr --export=__get_temp_ret --export=__set_temp_ret --export=__trap --export=__wasm_call_ctors --export=__get_exception_message --export=free --export=__thrown_object_from_unwind_exception --export=__cxa_increment_exception_refcount --export=__cxa_decrement_exception_refcount --export=malloc --export-table --growable-table -z stack-size=65536 --initial-memory=1073741824 --max-memory=4294967296 --no-entry --stack-first --table-base=1
 "/emsdk/upstream/bin/llvm-objcopy" out.wasm out.wasm --remove-section=producers
 "/emsdk/upstream/bin/wasm-emscripten-finalize" -g --dyncalls-i64 --pass-arg=legalize-js-interface-exported-helpers --dwarf out.wasm -o out.wasm --detect-features
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/src/compiler.js /tmp/tmp93zwfp28.json
emcc: warning: running limited binaryen optimizations because DWARF info requested (or indirectly required) [-Wlimited-postlink-optimizations]
 "/emsdk/upstream/bin/wasm-opt" --safe-heap out.wasm -o out.wasm -g --mvp-features --enable-exception-handling --enable-mutable-globals --enable-sign-ext
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp__ohs40f2/out.js unsignPointers --closureFriendly --exportES6 -o /tmp/emscripten_temp__ohs40f2/out.jso1.js
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp__ohs40f2/out.jso1.js safeHeap --closureFriendly --exportES6 -o /tmp/emscripten_temp__ohs40f2/out.jso2.js
root@a8c172869196:/mnt/linkerr# 
```

### Build under 3.1.48:
```
root@9b690697f719:/mnt/linkerr# emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.48 (e967e20b4727956a30592165a3c1cde5c67fa0a8)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

root@9b690697f719:/mnt/linkerr# make
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file1.o file1.cpp
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file2.o file2.cpp
  Linking 
 /emsdk/upstream/bin/clang --version
cache:INFO: generating system asset: symbol_lists/c80c1b80deac4f1d31b53986d27fdd418865c424.json... (this will be cached in "/emsdk/upstream/emscripten/cache/symbol_lists/c80c1b80deac4f1d31b53986d27fdd418865c424.json" for subsequent builds)
 /emsdk/node/16.20.0_64bit/bin/node /emsdk/upstream/emscripten/src/compiler.js /tmp/tmpizzavjf7.json --symbols-only
cache:INFO:  - ok
 /emsdk/upstream/bin/wasm-ld -o out.wasm ./file1.o ./file2.o -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten --whole-archive -lembind-rtti --no-whole-archive -lGL-webgl2-full_es3 -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc-debug -lcompiler_rt-wasm-sjlj -lc++-except -lc++abi-debug-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm /tmp/tmppk7ewi1ilibemscripten_js_symbols.so --export-if-defined=main --export-if-defined=__get_exception_message --export-if-defined=free --export-if-defined=__cpp_exception --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__thrown_object_from_unwind_exception --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=sbrk --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=emscripten_get_sbrk_ptr --export=__get_temp_ret --export=__set_temp_ret --export=__trap --export=__wasm_call_ctors --export=__get_exception_message --export=free --export=__thrown_object_from_unwind_exception --export=__cxa_increment_exception_refcount --export=__cxa_decrement_exception_refcount --export=malloc --export-table --growable-table -z stack-size=65536 --initial-memory=1073741824 --max-memory=4294967296 --no-entry --stack-first --table-base=1
wasm-ld: error: duplicate symbol: _emval_coro_resume
>>> defined in ./file1.o
>>> defined in ./file2.o
emcc: error: '/emsdk/upstream/bin/wasm-ld -o out.wasm ./file1.o ./file2.o -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten --whole-archive -lembind-rtti --no-whole-archive -lGL-webgl2-full_es3 -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc-debug -lcompiler_rt-wasm-sjlj -lc++-except -lc++abi-debug-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm /tmp/tmppk7ewi1ilibemscripten_js_symbols.so --export-if-defined=main --export-if-defined=__get_exception_message --export-if-defined=free --export-if-defined=__cpp_exception --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__thrown_object_from_unwind_exception --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=sbrk --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=emscripten_get_sbrk_ptr --export=__get_temp_ret --export=__set_temp_ret --export=__trap --export=__wasm_call_ctors --export=__get_exception_message --export=free --export=__thrown_object_from_unwind_exception --export=__cxa_increment_exception_refcount --export=__cxa_decrement_exception_refcount --export=malloc --export-table --growable-table -z stack-size=65536 --initial-memory=1073741824 --max-memory=4294967296 --no-entry --stack-first --table-base=1' failed (returned 1)
make: *** [makefile:76: all] Error 1
```

### file1.cpp
**Just remove the bind.h include and it will also link under 3.1.48.**
```cpp
#include <emscripten/emscripten.h>
#include <emscripten/bind.h>

EM_JS(void, file1testJS, (), {
    console.log ("file1testJS");
});

void file1test() {
    file1testJS();
}

extern void file2test();

int main() {
    file1test();
    file2test();
    return 0;
}
```

### file2.cpp
```cpp
#include <emscripten/emscripten.h>
#include <emscripten/bind.h>

EM_JS(void, file2testJS, (), {
    console.log ("file2testJS");
});

void file2test() {
    file2testJS();
}
```

--- a/system/include/emscripten/val.h
+++ b/system/include/emscripten/val.h
@@ -708,7 +708,7 @@ public:
 
 extern "C" {
   // JS FFI helper for `val_awaiter::resume_with`.
-  void _emval_coro_resume(val_awaiter* awaiter, EM_VAL result) {
+  inline void _emval_coro_resume(val_awaiter* awaiter, EM_VAL result) {
     awaiter->resume_with(val::take_ownership(result));
   }
 }

--- a/system/include/emscripten/val.h
+++ b/system/include/emscripten/val.h
@@ -708,7 +708,8 @@ public:
 
 extern "C" {
   // JS FFI helper for `val_awaiter::resume_with`.
-  inline void _emval_coro_resume(val_awaiter* awaiter, EM_VAL result) {
+  __attribute__((weak))
+  void _emval_coro_resume(val_awaiter* awaiter, EM_VAL result) {
     awaiter->resume_with(val::take_ownership(result));
   }
 }

--- a/test/test_other.py
+++ b/test/test_other.py
@@ -14162,3 +14162,11 @@ addToLibrary({
     # Test that even with `-Wl,--strip-all` the target features section is generated
     # by wasm-ld so that later phases (e.g. wasm-opt) can read it.
     self.do_runf('hello_world.c', emcc_args=['-Wl,--strip-all', '-pthread'])
+
+  def test_embind_no_duplicate_symbols(self):
+    # Embind implementation lives almost entirely in headers, which have special rules
+    # around symbol deduplication during linking. Ensure that including Embind headers
+    # in two different object files doesn't lead to linking errors.
+    create_file('a.cpp', '#include <emscripten/bind.h>')
+    create_file('b.cpp', '#include <emscripten/bind.h>')
+    self.run_process([EMXX, '-std=c++23', '-lembind', 'a.cpp', 'b.cpp'])

